src/common/multihash_isal.c

```diff
@@ -0,0 +1,444 @@
+/**
+ * (C) Copyright 2019-2020 Intel Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+ * The Government's rights to use, modify, reproduce, release, perform, display,
+ * or disclose this software are subject to the terms of the Apache License as
+ * provided in Contract No. B609815.
+ * Any reproduction of computer software, computer software documentation, or
+ * portions thereof marked with this legend must also reproduce the markings.
+ */
+
+#define D_LOGFAC	DD_FAC(csum)
+
+#include <stddef.h>
+#include <string.h>
+#include <stdlib.h>
+#include <stdint.h>
+
+#include <isa-l.h>
+#include <isa-l_crypto.h>
+#include <gurt/types.h>
+#include <daos/common.h>
+#include <daos/multihash.h>
+
+/**
+ * ---------------------------------------------------------------------------
+ * ISA-L Hash Algorithms
+ * ---------------------------------------------------------------------------
+ */
+
+/** CRC16_T10DIF*/
+static int
+crc16_init(void **daos_mhash_ctx)
+{
+	D_ALLOC(*daos_mhash_ctx, sizeof(uint16_t));
+	if (*daos_mhash_ctx == NULL)
+		return -DER_NOMEM;
+
+	return 0;
+}
+
+static int
+crc16_reset(void *daos_mhash_ctx)
+{
+	memset(daos_mhash_ctx, 0, sizeof(uint16_t));
+	return 0;
+}
+
+static void
+crc16_destroy(void *daos_mhash_ctx)
+{
+	D_FREE(daos_mhash_ctx);
+}
+
+static int
+crc16_update(void *daos_mhash_ctx, uint8_t *buf, size_t buf_len)
+{
+	uint16_t *crc16 = (uint16_t *)daos_mhash_ctx;
+
+	*crc16 = crc16_t10dif(*crc16, buf, (int)buf_len);
+	return 0;
+}
+
+static int
+crc16_finish(void *daos_mhash_ctx, uint8_t *buf, size_t buf_len)
+{
+	uint16_t *crc16 = (uint16_t *)daos_mhash_ctx;
+
+	*((uint16_t *)buf) = *crc16;
```

Comment:
@ryon-jensen Could you please advise whether we need to auto-reset on finish?

Reply:
For now it isn't expected that finish auto-resets, though maybe something we improve in the future.